uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -603,32 +768,55 @@ define([
     }
 
     /**
-     * Sets dates and types for vm.request from UI
+     * Sets dates (and times in case of "hours" absence type) for vm.request from UI
      */
     function _setDates () {
-      vm.request.from_date = vm.uiOptions.fromDate ? vm._convertDateToServerFormat(vm.uiOptions.fromDate) : null;
-      vm.request.to_date = vm.uiOptions.toDate ? vm._convertDateToServerFormat(vm.uiOptions.toDate) : null;
+      var options = vm.uiOptions;
+      var request = vm.request;
+      var times;
+
+      request.from_date = options.fromDate ? vm._convertDateToServerFormat(options.fromDate) : null;
+      request.to_date = options.toDate ? vm._convertDateToServerFormat(options.toDate) : null;
+
+      if (isCalculationUnit('hours') && !isLeaveType('toil')) {
+        times = options.times;
+        request.from_date = request.from_date && times.from.time ? request.from_date + ' ' + times.from.time : null;
+        request.to_date = request.to_date && times.to.time ? request.to_date + ' ' + times.to.time : null;
+        request.from_date_amount = !isNaN(+times.from.amount) ? times.from.amount : null;
+        request.to_date_amount = !isNaN(+times.to.amount) ? times.to.amount : null;
+      }
 
-      if (!vm.uiOptions.multipleDays && vm.uiOptions.fromDate) {
-        vm.uiOptions.toDate = vm.uiOptions.fromDate;
-        vm.request.to_date = vm.request.from_date;
+      if (!options.multipleDays && options.fromDate) {
+        options.toDate = options.fromDate;
+        request.to_date = request.from_date;
       }
     }
 
     /**
      * Sets dates and types for vm.request from UI
      */
     function _setDateAndTypes () {
+      var options = vm.uiOptions;
```

Comment:
**Comment:** This function is confusing. This is more about showing the balance than setting date and types. I see it's doing 3 things:

- sets the dates (by calling another function)
- determines if the balance is to be displayed
- updates the date type of the To field 

These balance displaying are heavily influenced by the multiple or single day options, but there is no mention of that in the description or the function name.

Your changes are good compared to what we previously had, but this function is in dire need of refactoring.

Reply:
I attempted to split this function into 2 functions.